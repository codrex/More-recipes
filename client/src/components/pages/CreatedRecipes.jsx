import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getCreatedRecipes } from '../../actions/recipeActions';
import RecipesDisplay from './RecipesDisplay';

/**
 * CreateRecipes component
 * @param {object} props
 * @return {React} react element
 */
const CreatedRecipes = props => (
  <RecipesDisplay
    title="Created recipes"
    getRecipes={props.getCreatedRecipes}
    recipes={props.createdRecipes}
    {...props}
  />
);

CreatedRecipes.propTypes = {
  getCreatedRecipes: PropTypes.func.isRequired,
  createdRecipes: PropTypes.arrayOf(PropTypes.shape).isRequired
};

const mapStateToProps = state => ({
  createdRecipes: state.recipes,
});

export { CreatedRecipes as PureCreatedRecipes };
export default connect(mapStateToProps, {
  getCreatedRecipes
})(CreatedRecipes);
