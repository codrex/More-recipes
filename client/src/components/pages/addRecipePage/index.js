import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Icon from '../../common/icon/icon';
import AddDirections from './addDirections/addDirections';
import AddIngredients from './addIngredients/addIngredients';
import RecipeNameAndCategory from './recipeNameAndCategory/recipeNameAndCategory';
import { createRecipe, updateAllRecipeField } from '../../../actions/recipeActions';
import { ajaxRedirect } from '../../../actions/ajaxActions';
import './addRecipe.scss';

/**
 * Add recipe component
 */
class AddRecipes extends React.Component {

  /**
   * @param {object} props
   */
  constructor(props) {
    super(props);
    this.saveRecipe = this.saveRecipe.bind(this);
  }
/**
 *
 * @param {object} nextProps
 * @return {bool} true or false
 */
  shouldComponentUpdate(nextProps) {
    if (nextProps.redirectUrl === '/') {
      this.props.actions.redirect('/recipe/create');
      this.props.history.push(nextProps.redirectUrl);
      return false;
    }
    if (nextProps.recipe !== this.props.recipe) {
      this.props.actions.updateLocalRecipe(nextProps.recipe);
      return true;
    }
    return true;
  }

  /**
   * @return {undefined}
   * @param {Object} values
   */
  saveRecipe(values) {
    this.props.actions.updateStore(values);
  }
  /**
   * @return {undefined}
   */
  render() {
    return (
      <div className="container-fluid no-padding ">
        <div className="row d-flex">
          <div className="col-xs-9 col-sm-9 col-md-10 col-lg-4 name-category no-padding">
            <h4 className="lead items-header-text">
              Add recipe
              <Icon
                iconClass="fa fa-plus"
                className="float-right"
                handleClick={this.newItem}
              />

            </h4>
            <div className="col-12">
              <RecipeNameAndCategory createRecipe={createRecipe} />
            </div>
          </div>
          <div className="col-xs-9 col-sm-9 col-md-10 col-lg-4 d-flex ingredients no-padding">
            <AddIngredients />
          </div>
          <div className="col-xs-9 col-sm-9 col-md-10 col-lg-4 d-flex directions no-padding">
            <AddDirections />
          </div>
        </div>
      </div>
    );
  }
}

AddRecipes.propTypes = {
  actions: PropTypes.object,
  loading: PropTypes.bool,
  redirectUrl: PropTypes.string,
  history: PropTypes.object,
  redirect: PropTypes.func,
  recipe: PropTypes.object,

};

const mapStateToProps = (state) => (
  {
    loading: state.ajaxCall > 0,
    redirectUrl: state.redirectUrl,
    recipe: state.recipe,
  }
);
const mapDispatchToProps = (dispatch) => (
  {
    actions: {
      redirect: bindActionCreators(ajaxRedirect, dispatch),
      updateLocalRecipe: bindActionCreators(updateAllRecipeField, dispatch)
    }
  }
);

export default connect(mapStateToProps, mapDispatchToProps)(AddRecipes);
