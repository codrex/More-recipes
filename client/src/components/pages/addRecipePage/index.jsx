import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import TopBar from '../../common/topbar/topbar';
import AddDirections from './addDirections/addDirections';
import AddIngredients from './addIngredients/addIngredients';
import RecipeNameAndCategory from './recipeNameAndCategory/recipeNameAndCategory';
import {
  updateAllRecipeField,
  recipeToModify,
  getRecipe,
  onNewRecipe
} from '../../../actions/recipeActions';
import { ajaxRedirect } from '../../../actions/ajaxActions';

/**
 * Add recipe component
 */
class AddRecipes extends React.Component {
  /**
   * @param {object} props
   */
  constructor(props) {
    super(props);
    this.create = props.match.path === '/recipe/create';
    this.CREATE_RECIPE_SUCCESS = 'Recipe added successfully';
    this.MODIFY_RECIPE_SUCCESS = 'Recipe modified successfully';
  }

  /**
 * @return {undefined}
 */
  componentDidMount() {
    if (!this.create) {
      const id = parseInt(this.props.match.params.id, 10);
      if (this.props.recipes) {
        const recipe = this.props.recipes.find(element => element.id === id);
        this.props.actions.recipeToModify(recipe);
      } else {
        this.props.actions.getRecipe(id);
      }
    } else {
      this.props.actions.newRecipe();
    }
  }
  /**
   * @return {undefined}
   * @param {Object} nextProps
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.success.success === this.CREATE_RECIPE_SUCCESS) {
      this.props.actions.newRecipe();
      this.props.actions.clearSuccessMsg({
        success: ''
      });
    }
  }

  /**
 *
 * @param {object} nextProps
 * @return {bool} true or false
 */
  shouldComponentUpdate(nextProps) {
    if (nextProps.redirectUrl === '/') {
      this.props.actions.redirect('/recipe/create');
      this.props.history.push(nextProps.redirectUrl);
      return false;
    }
    if (nextProps.recipe !== this.props.recipe) {
      this.props.actions.updateLocalRecipe(nextProps.recipe);
      return true;
    }
    return true;
  }

  /**
   * @return {undefined}
   */
  render() {
    return (
      <div
        className="container"
        style={{
          maxWidth: 'unset'
        }}
      >
        <div className="row">
          <TopBar title={this.create ? 'Add recipe' : 'Modify recipe'} />
        </div>
        <div
          className="row d-flex justify-content-between align-items-start"
          style={{
            margin: '0.5em'
          }}
        >
          <div className="col-xs-11 col-sm-11 col-md-10 col-lg-4 name-category no-padding">
            <div className="col-12">
              <RecipeNameAndCategory
                postRecipe={this.create}
                loading={this.props.loading}
                create={this.CREATE_RECIPE_SUCCESS}
                modify={this.MODIFY_RECIPE_SUCCESS}
              />
            </div>
          </div>
          <div className="col-xs-11 col-sm-11 col-md-10 col-lg-4 d-flex ingredients no-padding">
            <AddIngredients />
          </div>
          <div className="col-xs-11 col-sm-11 col-md-10 col-lg-4 d-flex directions no-padding">
            <AddDirections />
          </div>
        </div>
      </div>
    );
  }
}

AddRecipes.propTypes = {
  actions: PropTypes.object,
  loading: PropTypes.bool,
  redirectUrl: PropTypes.string,
  history: PropTypes.object,
  redirect: PropTypes.func,
  recipe: PropTypes.object,
  recipes: PropTypes.array,
  match: PropTypes.object
};

const mapStateToProps = state => ({
  loading: state.ajaxCall > 0,
  redirectUrl: state.redirectUrl,
  // recipes: state.user.createdRecipes,
  success: state.ajaxSuccess
});
const mapDispatchToProps = dispatch => ({
  actions: {
    redirect: bindActionCreators(ajaxRedirect, dispatch),
    updateLocalRecipe: bindActionCreators(updateAllRecipeField, dispatch),
    recipeToModify: bindActionCreators(recipeToModify, dispatch),
    getRecipe: bindActionCreators(getRecipe, dispatch),
    newRecipe: bindActionCreators(onNewRecipe, dispatch),
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(AddRecipes);
