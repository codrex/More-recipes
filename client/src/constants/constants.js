export const RECIPE_ADDED = 'Recipe was successfully posted';
export const RECIPE_MODIFIED = 'Recipe was successfully modified';
export const MODIFY_RECIPE_PATH = '/modify/:id';
export const CREATE_RECIPE_PATH = '/create';
export const CATEGORIES = [
  { value: 'breakfast', label: 'Breakfast' },
  { value: 'brunch', label: 'Brunch' },
  { value: 'lunch', label: 'Lunch' },
  { value: 'snacks', label: 'Bnacks' },
  { value: 'appertisers', label: 'Appertisers' },
  { value: 'dinner', label: 'Dinner' },
  { value: 'soup', label: 'Soup' },
  { value: 'noodles', label: 'Noodles' },
  { value: 'vegeterian', label: 'Vegeterian', },
  { value: 'salads', label: 'Salads' },
  { value: 'sides', label: 'Sides' },
  { value: 'sauces', label: 'Sauces' },
  { value: 'baking', label: 'Baking' },
  { value: 'desserts', label: 'Desserts' },
  { value: 'drinks', label: 'Drinks' },
  { value: 'cookies & candy', label: 'Cookies & Candy' },
  { value: 'miscellaneous', label: 'Miscellaneous' },
];
export const LIMIT = 10;
export const DEFAULT_PIX = 'https://res.cloudinary.com/resycom/image/upload/c_scale,q_53,w_2543/v1509718851/eaters-collective-132773_izkarh.jpg';
