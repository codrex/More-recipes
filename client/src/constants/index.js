/* eslint-disable max-len */
export const RECIPE_ADDED = 'Recipe was successfully posted';
export const RECIPE_MODIFIED = 'Recipe was successfully modified';
export const REVIEW_MESSAGE = 'Review posted';
export const CATEGORIES = [
  { value: 'breakfast', label: 'Breakfast' },
  { value: 'brunch', label: 'Brunch' },
  { value: 'lunch', label: 'Lunch' },
  { value: 'snacks', label: 'Snacks' },
  { value: 'appertisers', label: 'Appertisers' },
  { value: 'dinner', label: 'Dinner' },
  { value: 'soup', label: 'Soup' },
  { value: 'noodles', label: 'Noodles' },
  { value: 'vegeterian', label: 'Vegeterian', },
  { value: 'salads', label: 'Salads' },
  { value: 'sides', label: 'Sides' },
  { value: 'sauces', label: 'Sauces' },
  { value: 'baking', label: 'Baking' },
  { value: 'desserts', label: 'Desserts' },
  { value: 'drinks', label: 'Drinks' },
  { value: 'cookies and candy', label: 'Cookies and Candy' },
  { value: 'miscellaneous', label: 'Miscellaneous' },
];
export const LIMIT = 12;
export const REVIEW_LIMIT = 5;
export const DEFAULT_PIX = 'http://res.cloudinary.com/resycom/image/upload/c_scale,q_53,w_1073/v1515038728/photo-1510693206972-df098062cb71_gxgxnb.jpg';
export const DEFAULT_RECIPE_PIX = 'recipe-card-placeholder_ta9ikp_iebsf2.jpg';
export const DEFAULT_RECIPE_PIX_URL = 'https://res.cloudinary.com/resycom/image/upload/v1515305094/recipe-card-placeholder_ta9ikp_iebsf2.jpg';
export const PROFILE_PAGE_PIX = 'https://res.cloudinary.com/resycom/image/upload/c_scale,h_728,q_51/v1510430032/lily-lvnatikk-365344_nv94dc.jpg';
export const CAROUSEL_PIX_ONE = 'http://res.cloudinary.com/resycom/image/upload/c_scale,q_53,w_1073/v1515038728/photo-1510693206972-df098062cb71_gxgxnb.jpg';
export const CAROUSEL_PIX_TWO = 'https://res.cloudinary.com/resycom/image/upload/c_scale,q_53,w_2543/v1509718851/eaters-collective-132773_izkarh.jpg';
export const PRODUCTION_URL = 'https://more-recipesrex.herokuapp.com';
export const DEVELOPMENT_URL = 'http://127.0.0.1:8000';
export const PROTECTED_ROUTES = {
  Recipes: '/recipes',
  'Top Recipes': '/top-recipes',
  'Add Recipe': '/create'
};
export const UNPROTECTED_ROUTES = {
  login: '/login',
  'create account': '/create-account'
};
