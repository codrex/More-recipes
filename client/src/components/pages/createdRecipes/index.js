import CreatedRecipes from './createdRecipes';

export default CreatedRecipes;
