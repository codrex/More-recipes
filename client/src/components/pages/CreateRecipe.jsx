import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import RecipeEditor from './RecipeEditor';
import { resetSuccess } from '../../actions/ajaxActions';
import { RECIPE_ADDED } from '../../constants';
import {
  createRecipe,
  getRecipe,
  resetRecipe
} from '../../actions/recipeActions';

/**
 * CreateRecipe
 */
class CreateRecipe extends React.Component {
  /**
   * constructor
   * @return {undefined}
   * @param {Object} props
   */
  constructor(props) {
    super(props);
    props.resetRecipe();
  }

  /**
   * @returns {undefined}
   * @param {object} nextProps
   */
  componentWillReceiveProps(nextProps) {
    const {
      message,
      history,
      resetSuccessMessage
    } = nextProps;

    if (message === RECIPE_ADDED) {
      resetSuccessMessage();
      history.push(`/recipe/${nextProps.recipe.id}`);
    }
  }

  /**
   * post recipe button click event handler
   * @return {undefined}
   */
  postRecipeClicked = () => {
    const { recipe } = this.props;
    this.props.createRecipe(recipe, RECIPE_ADDED);
  }

  /**
   * render
   * @return {React} - react component
   */
  render() {
    const { loading } = this.props;

    return (
      <RecipeEditor
        handleButtonClick={this.postRecipeClicked}
        submitButtonText="post recipe"
        loading={loading}
        hasRender
      />
    );
  }
}

CreateRecipe.propTypes = {
  recipe: PropTypes.objectOf(PropTypes.shape).isRequired,
  history: PropTypes.objectOf(PropTypes.shape).isRequired,
  resetRecipe: PropTypes.func.isRequired,
  createRecipe: PropTypes.func.isRequired,
  resetSuccessMessage: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  message: PropTypes.string,
};

CreateRecipe.defaultProps = {
  message: '',
};

const mapStateToProps = state => (
  {
    recipe: state.recipe,
    loading: state.networkRequest.loading,
    message: state.networkRequest.msg,
    statusCode: state.currentStatusCode
  }
);

export { CreateRecipe as PureCreateRecipe };
export default connect(mapStateToProps, {
  resetRecipe,
  createRecipe,
  getRecipe,
  resetSuccessMessage: resetSuccess
})(CreateRecipe);
