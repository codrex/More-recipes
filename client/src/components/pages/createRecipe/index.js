import CreateRecipe from './createRecipe';

export default CreateRecipe;
