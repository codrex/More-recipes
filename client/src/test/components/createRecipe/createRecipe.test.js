/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import toJson from 'enzyme-to-json';
import { PureCreateRecipe } from '../../../components/pages/CreateRecipe';
import { RECIPE_ADDED } from '../../../constants';

const props = {
  resetRecipe: jest.fn(),
  message: '',
  getRecipe: jest.fn(),
  createRecipe: jest.fn(),
  resetSuccessMessage: jest.fn(),
  loading: false,
  statusCode: 200,
  recipe: {
    name: '',
    category: '',
    views: 0,
    upVotes: 0,
    downVotes: 0,
    image: {},
    id: 1
  },
  history: { push: jest.fn() },
};

describe('Create Recipe page component ', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<PureCreateRecipe {...props} />);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  test('expected render create recipe page', () => {
    const tree = toJson(wrapper);
    expect(wrapper.find('RecipeEditor').length).toBe(1);
    expect(tree).toMatchSnapshot();
    expect(tree).toBeInstanceOf(Object);
  });

  test(`should navigate to recipe details page when recipe is successfully
  created `, () => {
      wrapper.setProps({ message: RECIPE_ADDED });
      const tree = toJson(wrapper);
      expect(tree).toMatchSnapshot();
      expect(props.history.push).toHaveBeenCalledWith('/recipe/1');
      expect(tree).toBeInstanceOf(Object);
    });

  test('should post a recipe', () => {
    const tree = toJson(wrapper);
    wrapper.instance().postRecipe();
    expect(props.createRecipe).toHaveBeenCalledWith(props.recipe, RECIPE_ADDED);
    expect(tree).toMatchSnapshot();
    expect(tree).toBeInstanceOf(Object);
  });
});
