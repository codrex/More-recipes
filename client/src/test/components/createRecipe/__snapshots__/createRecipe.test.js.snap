// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Recipe page component ::  expected to match snapshot 1`] = `
<RecipeEditor
  handleButtonClick={[Function]}
  hasNotFound={false}
  hasRender={true}
  loading={false}
  message={[Function]}
  statusCode={-1}
  submitButtonText="post recipe"
/>
`;

exports[`Create Recipe page component ::  expected to match snapshot when recipe is created 1`] = `
<RecipeEditor
  handleButtonClick={[Function]}
  hasNotFound={false}
  hasRender={true}
  loading={false}
  message={[Function]}
  statusCode={-1}
  submitButtonText="post recipe"
/>
`;
